Add "How Clerk works" and "Client handshake" to initial concepts section (#577)
clerk/clerk-docs: docs/concepts/how-clerk-works.mdx

---
title: How Clerk works

> description: 

Reviewer: missing description

type: conceptual
---

# How Clerk works

## Development and production instances

> Production instances are used for production, pre-production, or staging environments with fixed domains. Production instances offer the strongest security and performance guarantees. For more details on Clerk internals please review the following sections.

Reviewer 2: ```suggestion
Production instances are used for production, pre-production, or staging environments with fixed domains. Production instances offer the strongest security and performance guarantees. Development instances contain certain trade-offs meant to support ease of active development. For more details on the differences, read the [Instances / Environments](/docs/deployments/environments) doc.
```

A user represents the current user of the session. The object holds all the basic user information e.g. name, email addresses, phone numbers, etc… including their public, private, and unsafe metadata.

### [Organization](/docs/organizations/roles-permissions)

An organization represents the current organization of the session. Users can belong to many organizations. One of them will be the current organization of the session. 

### [Roles](/docs/organizations/roles-permissions)

A user belongs to an organization with a role that defines their permissions. Currently, Clerk supports two roles, `org:admin` and `org:member`.

## Clerk Frontend API (FAPI)

Every Clerk development and production instance has a dedicated [Frontend API](https://clerk.com/docs/reference/frontend-api). This is the authentication, session, user & organization management API you or your users will interact with within the web browser, native application, Chrome extension, or Electron app.

Clerk assigns a unique random API URL such as `https://delicate-wahoo-73.accounts.dev/sign-in` to each development instance. For production instances, the Frontend API lives at `https://clerk.[your-domain.com]`.

> ### How it works

Reviewer: ```suggestion
### How Clerk's Frontend API works
```

Clerk's Frontend API is responsible for maintaining the authentication state of the current client (browser or native application). Based on the authentication's state (`signed-in`, `signed-out`, `handshake`) it mints a short-lived session [JSON Web Token (JWT)](/docs/backend-requests/making/jwt-templates) for the current user and session to be used by the host application that is secured by Clerk. The session JWT is stored in a cookie or can be retrieved at any time via the [`getToken()`](/docs/references/javascript/session#get-token) method.

### Clerk cookies

Clerk’s Frontend API uses two cookies for session management in production instances:

* `__client` cookie: a long-lived session that is used to keep the the session JWT fresh by interacting with the Frontend API.
* `__session` cookie: a short-lived session JWT to validate requests on your application or your API. 

#### The `__client` cookie

> This cookie is set on Clerk Frontend API (for example `https://clerk.<example.com>`) for each instance. It’s HTTP-only, first-party, and secure. It contains a long-lived client JWT that lasts 7 days by default. The duration is configurable in the Clerk Dashboard's  [Sessions page](https://dashboard.clerk.com/last-active?path=sessions). The JWT identifies the current client (browser, native application, or chrome extension) and sets the current active sessions.

Reviewer: ```suggestion
The `__client` cookie is set on Clerk Frontend API (for example `https://clerk.<example.com>`) for each instance. It’s HTTP-only, first-party, and secure. It contains a long-lived client JWT that lasts 7 days by default. The duration is configurable in the Clerk Dashboard's  [Sessions page](https://dashboard.clerk.com/last-active?path=sessions). The JWT identifies the current client (browser, native application, or chrome extension) and sets the current active sessions.
```

#### The `__session` cookie

> This cookie is set on the host application that is secured by Clerk (for example `https://dashboard.<example.com>`). It’s a JS cookie, it’s secure, and contains a short-lived session JWT that lasts 60 seconds. This JWT contains the current session, user, and organization identifiers and must be sent to the application API to authenticate API requests. In SSR, the `__session` cookie travels automatically to the server. ClerkJS ensures that the `__session` cookie is automatically refreshed in Client Side Rendering (CSR) and Server Side Rendering (SSR), ensuring a fresh session is always available.

Reviewer: ```suggestion
The `__session` cookie is set on the host application that is secured by Clerk (for example `https://dashboard.<example.com>`). It’s a JS cookie, it’s secure, and it contains a short-lived session JWT that lasts 60 seconds. This JWT contains the current session, user, and organization identifiers and must be sent to the application API to authenticate API requests. In SSR, the `__session` cookie travels automatically to the server. ClerkJS ensures that the `__session` cookie is automatically refreshed in Client Side Rendering (CSR) and Server Side Rendering (SSR), ensuring a fresh session is always available.
```

However, if a `__session` cookie is expired on a request to an application’s backend, the SDK doesn’t know if the session has ended, or if a new short-lived JWT needs to be issued. When an SDK gets into this state, it triggers Clerk's client handshake.

### The client handshake

Clerk uses a client handshake mechanism to resolve a request’s authentication state from unknown (`handshake`) to `signed-in` or `signed-out`. 

With this handshake, Clerk can resolve authentication state on the backend and ensure the request is properly handled as signed in or out, instead of being in a potentially unknown state. The handshake flow relies on redirects to exchange session information between FAPI and the application, ensuring the resolution of unknown authentication states minimizes performance impact and behaves consistently across different framework and language implementations.

> Learn more about [how Clerk's client handshake works in this deep dive](/docs/concepts/client-handshake).

Reviewer: ```suggestion
Learn more about [how Clerk's client handshake works.](/docs/concepts/client-handshake).
```

### Client piggybacking

> Every Frontend API request except from `/v1/client/sessions/:id/tokens` that refreshes the Clerk session tokens, returns a response payload that contains two top level keys

Reviewer: We can move this information into a callout. See the comment a few below this one.

```suggestion
Every Frontend API request returns a response payload that contains two top level keys: 
```

```json

> {
>    client: ...,
>    response: ...
> }

Reviewer: ```suggestion
{
   "client": ...,
   "response": ...
}
```

```

> The `response` key contains the resource or resources of each endpoint. The `client` key contains the client piggybacking payload. The client object identifies the current device and contains all sessions, users and organization data for the current device (see [Main Objects](#main-objects) section). 

Reviewer: I think this information can be thrown in a callout, because as its currently written above, its interrupting that sentence and makes the information confusing.

```suggestion
<Callout type="info">
  The `/v1/client/sessions/:id/tokens` endpoint refreshes the Clerk session tokens and is an exception to this format.
</Callout>

The `response` key contains the resource or resources of each endpoint. The `client` key contains the client piggybacking payload. The client object identifies the current device and contains all sessions, users, and organization data for the current device (see [Main Objects](#main-objects) section). 
```

> This is effectively the necessary frontend state for ClerkJS and powers the JS SDKs, the React hooks and the Components. It is a best-effort way for the backend to update the frontend state on every Frontend API request. 

Reviewer: ```suggestion
This is effectively the necessary frontend state for ClerkJS and powers the JS SDKs, the [React hooks](/docs/references/react/use-user), and [Clerk components](/docs/components/overview). It is a best-effort way for the backend to update the frontend state on every Frontend API request. 
```

It's not recommended to use Client piggybacking because of the following issues:

1. It’s a very large chunk of a tree-like state and doesn’t change often.

> 1. It’s hard to compute on the backend as it requires a lot of data from the database. The computation takes place on every request adding a significant overhead.

Reviewer: ```suggestion
1. It’s hard to compute on the backend as it requires a lot of data from the database. The computation takes place on every request, adding a significant overhead.
```

---
title: How Clerk works
description: Learn about the main concepts and objects that power Clerk's powerful authentication and user management system.
type: conceptual
---

> # How Clerk works

Reviewer 3: It might be good to link to our `How We Roll` blog series here as well: https://clerk.com/blog/how-we-roll-roundup

#### The `__session` cookie

The `__session` cookie is set on the host application that is secured by Clerk (for example `https://dashboard.<example.com>`). It’s a JS cookie, it’s secure, and it contains a short-lived session JWT that lasts 60 seconds. This JWT contains the current session, user, and organization identifiers and must be sent to the application API to authenticate API requests. In SSR, the `__session` cookie travels automatically to the server. ClerkJS ensures that the `__session` cookie is automatically refreshed in Client Side Rendering (CSR) and Server Side Rendering (SSR), ensuring a fresh session is always available.

However, if a `__session` cookie is expired on a request to an application’s backend, the SDK doesn’t know if the session has ended, or if a new short-lived JWT needs to be issued. When an SDK gets into this state, it triggers Clerk's client handshake.

> 

Reviewer 4: #### The `__client_uat` cookie

This `__client_uat` cookie is set on the host application that is secured by Clerk and hints Clerk SDKs about the timestamp of the latest authentication update of the current client.

Reviewer 4: @alexisintech Please add this ☝️

```json
{
   "client": ...,
   "response": ...
}
```

<Callout type="info">
  The `/v1/client/sessions/:id/tokens` endpoint refreshes the Clerk session tokens and is an exception to this format.
</Callout>

The `response` key contains the resource or resources of each endpoint. The `client` key contains the client piggybacking payload. The client object identifies the current device and contains all sessions, users, and organization data for the current device (see [Main Objects](#main-objects) section). 

This is effectively the necessary frontend state for ClerkJS and powers the JS SDKs, the [React hooks](/docs/references/react/use-user), and [Clerk components](/docs/components/overview). It is a best-effort way for the backend to update the frontend state on every Frontend API request. 

It's not recommended to use Client piggybacking because of the following issues:

1. It’s a very large chunk of a tree-like state and doesn’t change often.
1. It’s hard to compute on the backend as it requires a lot of data from the database. The computation takes place on every request, adding a significant overhead.
1. It’s best effort.

> 1. It’s not a two-way communication channel.

Reviewer 3: This section feels odd here. Who is the audience? Users interested in using the data? Clerk employees? Why are we documenting it if we don't want it to be used?

Reviewer 4: Please move this under the Clerk Frontend API section.

Reviewer 4: @alexisintech ☝️

Reviewer: Is this not the section you meant? Because this is where I moved it to:

<img width="820" alt="Screenshot 2024-02-05 at 10 28 17" src="https://github.com/clerk/clerk-docs/assets/98043211/0a3d5990-80cc-4c41-8f6c-eb7998180863">

> ### Clerk Frontend API for non-standard web browsers

Reviewer 4: Please move this under the Clerk Frontend API section.

Reviewer 4: Also add a link to https://clerk.com/docs/reference/frontend-api similarly to BAPI.
